### PR TITLE
RUBY-1174 RUBY-1178 Update Max Staleness implementation

### PR DESCRIPTION
**Pull request by estolfo**

### CODE REVIEW

**Comment by estolfo** [comment]
@jyemin and @rozza, it'd be great if you could review this. Thanks!

File: lib/mongo/server_selector/selectable.rb [hunk 1]
```diff
@@ -154,8 +154,9 @@ def candidates(cluster)
         elsif cluster.sharded?
           near_servers(cluster.servers).each { |server| validate_max_staleness_support!(server) }
         else
-          validate_max_staleness_value!(cluster)
-          select(cluster.servers)
+          servers = cluster.servers
+          validate_max_staleness_value!(cluster) unless servers.empty?
```
Comment (estolfo) [inline lib/mongo/server_selector/selectable.rb]: ReplicaSetNoPrimary/NoKnownServers.yml was added after my day ended yesterday so was not in this PR.
I've just updated them to the latest, was failing ReplicaSetNoPrimary/NoKnownServers.yml but then changed the code to do the check unless Topology type is Unknown.

File: lib/mongo/server_selector/selectable.rb [hunk 2]
```diff
@@ -258,10 +259,13 @@ def validate_max_staleness_support!(server)
       end
 
       def validate_max_staleness_value!(cluster)
-        return unless @max_staleness
-        heartbeat_frequency = cluster.options[:heartbeat_frequency] || Server::Monitor::HEARTBEAT_FREQUENCY
-        if @max_staleness < heartbeat_frequency * 2
-          raise Error::InvalidServerPreference.new(Error::InvalidServerPreference::INVALID_MAX_STALENESS)
+        if @max_staleness
+          heartbeat_frequency = cluster.options[:heartbeat_frequency] || Server::Monitor::HEARTBEAT_FREQUENCY
```
Comment (estolfo) [inline lib/mongo/server_selector/selectable.rb]: done

File: lib/mongo/server_selector/selectable.rb [hunk 3]
```diff
@@ -258,10 +259,13 @@ def validate_max_staleness_support!(server)
...
+          unless @max_staleness > 0  &&
+                   @max_staleness >= [ SMALLEST_MAX_STALENESS_SECONDS,
+                                       (heartbeat_frequency  + Cluster::IDLE_WRITE_PERIOD_SECONDS) ].max
+            raise Error::InvalidServerPreference.new(Error::InvalidServerPreference::INVALID_MAX_STALENESS)
```
Comment (estolfo) [inline lib/mongo/server_selector/selectable.rb]: done

**Comment by estolfo** [comment]
@jyemin, I've addressed all comments. If you could review one more time, that'd be great.

**Comment by jyemin** [comment]
LGTM

**Comment by estolfo** [comment]
@jarthod Do you want to take a look at this as well to verify unit conversions?

**Comment by jarthod** [comment]
Looks good to me too ;)

**Comment by estolfo** [comment]
thanks @jarthod = )

**Pull request merged**